# Main entrypoint of the workflow.
# Please follow the best practices:
# https://snakemake.readthedocs.io/en/stable/snakefiles/best_practices.html,
# in particular regarding the standardized folder structure mentioned there.
import pandas as pd


configfile: "config/config.yaml"


include: "rules/ref.smk"
include: "rules/download.smk"
include: "rules/trim.smk"
include: "rules/mapping.smk"
include: "rules/downsampling.smk"
include: "rules/contaminate.smk"


rule all:
    input:
        expand(
            "results/contaminated/{tumor_sample}.{chr}.{total_depth}x.{contamination}_percent_{normal_sample}_contamination.bam",
            tumor_sample=config["pure_tumor_sample"]["sample_name"],
            chr=config.get("chromosome_to_extract", "full_genome"),
            total_depth=config["target_coverages"],
            contamination=config["target_normal_contaminations_percent"],
            normal_sample=config["pure_normal_sample"]["sample_name"],
        ),
        expand(
            "results/downsampled/{normal_sample}.{chr}.{depth}x.bam",
            normal_sample=config["pure_normal_sample"]["sample_name"],
            chr=config.get("chromosome_to_extract", "full_genome"),
            depth=config["target_coverages"],
        ),


wildcard_constraints:
    sample=f"{config['pure_tumor_sample']['sample_name']}|{config['pure_normal_sample']['sample_name']}",
    chr=f"{config['chromosome_to_extract']}",
